### PR TITLE
Add packages that support parsing and writing new attribution codes to DMGs

### PR DESCRIPTION
**Pull request by bhearsum**
This is heavy lifting for what's needed for #147 - a Go implementation of DMG parsing & updating of attribution codes. I've tried to isolate all of the core DMG parsing and updating code in the new `dmglib` package, and keep the specific flow of writing new attribution codes in `dmgmodify`.

I've included a small attributable DMG that's used in one of the tests, but also can be used to test by hand with commands like:
```
go run dmgmodify/main.go dmglib/tests/fixtures/attributable.dmg /tmp/attributed.dmg __MOZCUSTOM__ __MOZCUSTOM__with_some_attribution_data
```

You can then verify that DMG with:
```
hdiutil verify attributed.dmg
hdiutil attach attributed.dmg
xattr -p com.apple.application-instance /Volume/untitled/a
```

The `hdiutil verify` should report valid checksums, and you should see the updated attribution string from `xattr`.

### CODE REVIEW

File: dmgmodify/dmgmodify/modify.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	ErrCodeTooLong            = errors.New("dmgmodify: attribution code is too long")
+	TAB                       = 0x9
+	NUL                       = 0x0
+	crcPolynomial      uint32 = 0xedb88320
+)
+
+// Update `dmg`, replacing the `sentinel` area with the provided `code`.
+// This function is a port of the C implementation from libdmg-hfsplus
+// (https://github.com/mozilla/libdmg-hfsplus/blob/a0a959bd25370c1c0a00c9ec525e3e78285adbf9/dmg/attribution.c#L209)
+// Note: We explicitly do _not_ update the Attribution resource here, as it
```
Comment (willdurand) [inline dmgmodify/dmgmodify/modify.go:27]: nit:
```suggestion
// Note: We explicitly do _not_ update the Attribution resource here, as it is
```

File: dmgmodify/dmgmodify/modify.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+// not necessary, and makes for unnecessary work in the critical path of a
+// new Firefox install. This does not impact the attribution of the build, but
+// it does mean the build cannot be re-attributed later (which is not something
+// that ever needs to happen).
+func WriteAttributionCode(dmg *dmglib.DMG, sentinel string, code []byte) error {
+	// First, pull the information we need to update the attribution code.
+	// The blkx resource has some metadata that we need to update after
+	// injecting the attribution code.
+	blkxRes, _ := dmg.Resources.GetResourceDataByName("blkx")
```
Comment (willdurand) [inline dmgmodify/dmgmodify/modify.go:36]: I am wondering why we silently swallow the error here? Compared to, e.g.:

```suggestion
	blkxRes, err := dmg.Resources.GetResourceDataByName("blkx")
	if err != nil {
	  return err
	}
```

File: dmgmodify/dmgmodify/modify.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	// The plst resource contains some metadata that help us quickly
+	// locate the attribution data, and update the blkx and top level dmg
+	// metadata.
+	plstRes, _ := dmg.Resources.GetResourceDataByName("plst")
```
Comment (willdurand) [inline dmgmodify/dmgmodify/modify.go:40]: ditto

File: dmgmodify/main.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	output, err := os.Create(os.Args[2])
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer output.Close()
+
+	dmgObj, err := input.Parse()
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (willdurand) [inline dmgmodify/main.go:30]: Shouldn't we (defer) close the `dmgObj` here too?

File: dmglib/types.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package dmglib
+
+import (
+	"io"
+)
+
+type ReaderSeeker interface {
```
Comment (willdurand) [inline dmglib/types.go:7]: It'd be good to document these types. IIRC these are used to "ease" testing.

That said, I am wondering what the `ReaderSeeker` is. The stdlib already provides `ReadSeeker`: https://pkg.go.dev/io#ReadSeeker

File: dmglib/types.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+	io.Reader
+	io.ReaderAt
+	io.Seeker
+}
+
+type ReaderAtSeeker interface {
```
Comment (willdurand) [inline dmglib/types.go:13]:
```suggestion

// ReadAtSeeker is the interface that groups the basic ReadAt and Seek methods.
type ReaderAtSeeker interface {
```

**Review by willdurand — COMMENTED** [review]
Thanks, and sorry for the delay. I have some questions/comments, and I am concerned about the missing coverage.

File: dmgmodify/dmgmodify/modify.go (hunk 2 above)
Comment (bhearsum) [reply dmgmodify/dmgmodify/modify.go:36]: No good reason I can think of...I probably forgot to update this after some early testing/hacking.

File: dmglib/types.go (hunk 5 above)
Comment (bhearsum) [reply dmglib/types.go:7]: I'll document them. It looks like there's no compound types in the stdlib that include `ReaderAt`?

**Comment by bhearsum** [comment]
> Thanks, and sorry for the delay. I have some questions/comments, and I am concerned about the missing coverage.

I'll work on improving the coverage. Are there specific parts you're more concerned with than others? The main reason it's so low is because I wasn't sure we wanted to check in opaque DMGs, and the necessary data is too large to embed in the test files.

File: dmgmodify/main.go (hunk 4 above)
Comment (bhearsum) [reply dmgmodify/main.go:30]: Actually...no. This is a `DMG` object that has no file handles associated with it.

**Comment by willdurand** [comment]
> Are there specific parts you're more concerned with than others? The main reason it's so low is because I wasn't sure we wanted to check in opaque DMGs, and the necessary data is too large to embed in the test files.

It's more a general comment like "maybe we should find a way to cover those lines" after I saw the different codecov comments. We don't update this project frequently so preventing all kind of regressions by having excellent coverage would be great, ideally. I am not going to block on that, though.

**Comment by willdurand** [comment]
> Are there specific parts you're more concerned with than others? The main reason it's so low is because I wasn't sure we wanted to check in opaque DMGs, and the necessary data is too large to embed in the test files.

It's more a general comment like "maybe we should find a way to cover those lines" after I saw the different codecov comments. We don't update this project frequently so preventing all kind of regressions by having excellent coverage would be great, ideally. I am not going to block on that, though.

**Comment by bhearsum** [comment]
> > Are there specific parts you're more concerned with than others? The main reason it's so low is because I wasn't sure we wanted to check in opaque DMGs, and the necessary data is too large to embed in the test files.
> 
> It's more a general comment like "maybe we should find a way to cover those lines" after I saw the different codecov comments. We don't update this project frequently so preventing all kind of regressions by having excellent coverage would be great, ideally. I am not going to block on that, though.

That's fair. I improved the test coverage a fair bit - it now covers the entire happy path, and a little bit more of the error handling. (Some of the uncovered error lines are not worth testing IMO, eg: `binary.Encode` can only return EOFs.)

**Comment by bhearsum** [comment]
I integrated all of my changes with the existing commits. Here's a compare link with just the new changes (tip of what you reviewed previously vs. new tip): https://github.com/mozilla-services/stubattribution/compare/21d2f72ad56383f1c5cfbe13e7af7e489fa46010..105cd685b7928a8c40aeb727c6c1b6902877dd11 if you want it.

**Review by willdurand — APPROVED** [review]

**Pull request merged**